pkg/flexvolume/flexvolume.go

```diff
@@ -141,129 +141,232 @@ func processOpts(optsStr string) (Options, error) {
 	case "waitforattach":
 		if len(args) != 4 {
-			ExitWithResult(Fail("waitforattach expected exactly 4 arguments; got ", args))
+			return Fail("waitforattach expected exactly 4 arguments; got ", args)
 		}
 
 		mountDevice := args[2]
 		opts, err := processOpts(args[3])
 		if err != nil {
-			ExitWithResult(Fail(err))
+			return Fail(err.Error())
+		}
+		volumeID, ok := opts[OptionKeypvOrVolumeName]
+		if !ok {
+			return Fail("key '%s' not found in attach options",
+				OptionKeypvOrVolumeName)
 		}
 
-		ExitWithResult(driver.WaitForAttach(mountDevice, opts))
+		driver := ClaimVolume(volumeID, drivers)
+		if driver == nil {
+			return Fail("Unable to find flexdriver for volume id '%s'", volumeID)
+		}
+
+		return driver.WaitForAttach(mountDevice, opts)
 
 	// <driver executable> isattached <json options> <node name>
 	case "isattached":
 		if len(args) != 4 {
-			ExitWithResult(Fail("isattached expected exactly 4 arguments; got ", args))
+			return Fail("isattached expected exactly 4 arguments; got ", args)
 		}
 
 		opts, err := processOpts(args[2])
 		if err != nil {
-			ExitWithResult(Fail(err))
+			return Fail(err.Error())
+		}
+		volumeID, ok := opts[OptionKeypvOrVolumeName]
+		if !ok {
+			return Fail("key '%s' not found in attach options:%#v",
+				OptionKeypvOrVolumeName,
+				opts)
 		}
+
 		nodeName := args[3]
-		ExitWithResult(driver.IsAttached(opts, nodeName))
+		driver := ClaimVolume(volumeID, drivers)
+		if driver == nil {
+			return Fail("Unable to find flexdriver for volume id '%s'", volumeID)
+		}
+
+		return driver.IsAttached(opts, nodeName)
 
 	// <driver executable> mountdevice <mount dir> <mount device> <json options>
 	case "mountdevice":
 		if len(args) != 5 {
-			ExitWithResult(Fail("mountdevice expected exactly 5 arguments; got ", args))
+			return Fail("mountdevice expected exactly 5 arguments; got ", args)
 		}
 
 		mountDir := args[2]
 		mountDevice := args[3]
 
 		opts, err := processOpts(args[4])
 		if err != nil {
-			ExitWithResult(Fail(err))
+			return Fail(err.Error())
 		}
 
-		ExitWithResult(driver.MountDevice(mountDir, mountDevice, opts))
+		volumeID, ok := opts[OptionKeypvOrVolumeName]
+		if !ok {
+			return Fail("key '%s' not found in attach options",
+				OptionKeypvOrVolumeName)
+		}
+
+		driver := ClaimVolume(volumeID, drivers)
+		if driver == nil {
+			return Fail("Unable to find flexdriver for volume id '%s'", volumeID)
+		}
+
+		return driver.MountDevice(mountDir, mountDevice, opts)
 
 	// <driver executable> unmountdevice <mount dir>
 	case "unmountdevice":
 		if len(args) != 3 {
-			ExitWithResult(Fail("unmountdevice expected exactly 3 arguments; got ", args))
+			return Fail("unmountdevice expected exactly 3 arguments; got ", args)
 		}
 
 		mountDir := args[2]
-		ExitWithResult(driver.UnmountDevice(mountDir))
+
+		// assumes that the last part of mountDir is the volumeID
+		parts := strings.Split(mountDir, "/")
+		volumeID := parts[len(parts)-1]
```

Comment:
Maybe we should do some bounds checking here, just to be on the safe side? i.e. if someone passed in an invalid mountDir, would this panic?